src/java.base/share/classes/java/lang/invoke/StringConcatFactory.java

```diff
@@ -1191,24 +1192,36 @@ private static MethodTypeDesc prependArgs(MethodType concatArgs) {
                 var cl = concatArgs.parameterType(i);
                 paramTypes[i + 4] = needStringOf(cl) ? CD_String : ConstantUtils.classDesc(cl);
             }
-            return MethodTypeDesc.of(CD_int, paramTypes);
+            return MethodTypeDescImpl.ofValidated(CD_int, paramTypes);
         }
 
         /**
-         * Construct the MethodType of the coder method,
+         * Construct the MethodType of the coder method, if there are no parameters it may be UTF16, return null.
          * The first parameter is the initialized coder, Only parameter types that can be UTF16 are added.
```

Comment:
```suggestion
         * Construct the MethodType of the coder method. The first parameter is the initialized coder.
         * Only parameter types which can be UTF16 are added. Returns null if no such parameter exists.
```

Reply:
Something to consider (for a follow up) is to not add the initial coder as an argument but test that in the outer method: No point calling the `coder` method if we already know that we're UTF-16.

Reply:
If initCoder is UTF16, the coder method does not need to be generated. This is an optimization and may be a separate PR. Of course, if you agree, I can also put it here.

Reply:
Right, if `initCoder` is always UTF16 (`-XX:-CompactStrings`) then we never need the `coder` method. Nor really if any of the constants have UTF-16 chars, but in those cases it might be more appropriate to inject the derived value and have a runtime test - as long as it get optimized away. Otherwise we might need to generate two different classes per shape.

Reply:
If we're to pre-generate code for concats when linking JDK images (something that might be a big help for startup) then it'd be nice to be agnostic regarding `CompactStrings`